core-primitives/attestation-handler/src/attestation_handler.rs

```diff
@@ -0,0 +1,638 @@
+//    contributors may be used to endorse or promote products derived
+//    from this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+// "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+// LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+// A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+// OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+// LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+// DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+// THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+use crate::{cert, utils::hash_from_slice, Error as EnclaveError, Result as EnclaveResult};
+use codec::Encode;
+use core::default::Default;
+use itertools::Itertools;
+use itp_node_api::{
+	api_client::{ParentchainExtrinsicParams, ParentchainExtrinsicParamsBuilder},
+	metadata::{pallet_teerex::TeerexCallIndexes, provider::AccessNodeMetadata, NodeMetadata},
+};
+use itp_ocall_api::EnclaveAttestationOCallApi;
+use itp_settings::files::{RA_API_KEY_FILE, RA_DUMP_CERT_DER_FILE, RA_SPID_FILE};
+use itp_sgx_crypto::Ed25519Seal;
+use itp_sgx_io as io;
+use itp_sgx_io::StaticSealedIO;
+use itp_utils::write_slice_and_whitespace_pad;
+use log::*;
+use sgx_rand::*;
+use sgx_tcrypto::*;
+use sgx_tse::*;
+use sgx_types::*;
+use sp_core::{blake2_256, Pair};
+use std::{
+	format,
+	io::{Read, Write},
+	net::TcpStream,
+	prelude::v1::*,
+	println, slice, str,
+	string::String,
+	sync::Arc,
+	vec::Vec,
+};
+use substrate_api_client::{compose_extrinsic_offline, ExtrinsicParams};
+
+pub const DEV_HOSTNAME: &str = "api.trustedservices.intel.com";
+
+#[cfg(feature = "production")]
+pub const SIGRL_SUFFIX: &str = "/sgx/attestation/v4/sigrl/";
+#[cfg(feature = "production")]
+pub const REPORT_SUFFIX: &str = "/sgx/attestation/v4/report";
+
+#[cfg(not(feature = "production"))]
+pub const SIGRL_SUFFIX: &str = "/sgx/dev/attestation/v4/sigrl/";
+#[cfg(not(feature = "production"))]
+pub const REPORT_SUFFIX: &str = "/sgx/dev/attestation/v4/report";
+
+pub struct AttestationHandler<OCallApi, NodeMetadataRepository> {
+	ocall_api: Arc<OCallApi>,
+	node_metadata_repo: Arc<NodeMetadataRepository>,
+}
+
+impl<OCallApi, NodeMetadataRepository> AttestationHandler<OCallApi, NodeMetadataRepository>
+where
+	OCallApi: EnclaveAttestationOCallApi,
+	NodeMetadataRepository: AccessNodeMetadata<MetadataType = NodeMetadata>,
+{
+	pub fn new(ocall_api: Arc<OCallApi>, node_metadata_repo: Arc<NodeMetadataRepository>) -> Self {
+		Self { ocall_api, node_metadata_repo }
+	}
+
+	pub unsafe fn get_mrenclave(&self, mrenclave: *mut u8, mrenclave_size: u32) -> sgx_status_t {
```

Comment:
I think we can also refactor this, from a legacy C-interface to a proper Rust function interface, returning a Result<MrEnclave>, instead of an unsafe handling of raw pointers?